[Workflow] Introducing the workflow component

lyrixx opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | not yet |
| Fixed tickets | n/a |
| License | MIT |
| Doc PR | n/a |

TODO:
- [x] Add tests
- [x] Add PHP doc
- [x] Add Symfony fullstack integration (Config, DIC, command to dump the state-machine into graphiz format)

So why another component?

This component take another approach that what you can find on [Packagist](https://packagist.org/search/?q=state%20machine).

Here, the workflow component is not tied to a specific object like with [Finite](https://github.com/yohang/Finite). It means that the component workflow is stateless and can be a symfony service.

Some code:

``` php
#!/usr/bin/env php
<?php

require __DIR__.'/vendor/autoload.php';

use Symfony\Component\EventDispatcher\Debug\TraceableEventDispatcher;
use Symfony\Component\EventDispatcher\EventDispatcher;
use Symfony\Component\Stopwatch\Stopwatch;
use Symfony\Component\Workflow\Definition;
use Symfony\Component\Workflow\Dumper\GraphvizDumper;
use Symfony\Component\Workflow\Marking;
use Symfony\Component\Workflow\MarkingStore\PropertyAccessorMarkingStore;
use Symfony\Component\Workflow\MarkingStore\ScalarMarkingStore;
use Symfony\Component\Workflow\Transition;
use Symfony\Component\Workflow\Workflow;

class Foo
{
    public $marking;

    public function __construct($init = 'a')
    {
        $this->marking = $init;
        $this->marking = [$init => 1];
    }
}

$fooDefinition = new Definition(Foo::class);
$fooDefinition->addPlaces([
    'a', 'b', 'c', 'd', 'e', 'f', 'g',
]);

//                                           name  from        to
$fooDefinition->addTransition(new Transition('t1', 'a',        ['b', 'c']));
$fooDefinition->addTransition(new Transition('t2', ['b', 'c'],  'd'));
$fooDefinition->addTransition(new Transition('t3', 'd',         'e'));
$fooDefinition->addTransition(new Transition('t4', 'd',         'f'));
$fooDefinition->addTransition(new Transition('t5', 'e',         'g'));
$fooDefinition->addTransition(new Transition('t6', 'f',         'g'));

$graph = (new GraphvizDumper())->dump($fooDefinition);

$ed = new TraceableEventDispatcher(new EventDispatcher(), new Stopwatch(), new \Monolog\Logger('app'));

// $workflow = new Workflow($fooDefinition, new ScalarMarkingStore(), $ed);
$workflow = new Workflow($fooDefinition, new PropertyAccessorMarkingStore(), $ed);

$foo = new Foo(isset($argv[1]) ? $argv[1] : 'a');

$graph = (new GraphvizDumper())->dump($fooDefinition, $workflow->getMarking($foo));

dump([
    'AvailableTransitions' => $workflow->getAvailableTransitions($foo),
    'CurrentMarking' => clone $workflow->getMarking($foo),
    'can validate t1' => $workflow->can($foo, 't1'),
    'can validate t3' => $workflow->can($foo, 't3'),
    'can validate t6' => $workflow->can($foo, 't6'),
    'apply t1' => clone $workflow->apply($foo, 't1'),
    'can validate t2' => $workflow->can($foo, 't2'),
    'apply t2' => clone $workflow->apply($foo, 't2'),
    'can validate t1 bis' => $workflow->can($foo, 't1'),
    'can validate t3 bis' => $workflow->can($foo, 't3'),
    'can validate t6 bis' => $workflow->can($foo, 't6'),
]);

```

The workflown:

![workflow](https://cloud.githubusercontent.com/assets/408368/14183999/4a43483c-f773-11e5-9c8b-7f157e0cb75f.png)

The output:

```
array:10 [
  "AvailableTransitions" => array:1 [
    0 => Symfony\Component\Workflow\Transition {#4
      -name: "t1"
      -froms: array:1 [
        0 => "a"
      ]
      -tos: array:2 [
        0 => "b"
        1 => "c"
      ]
    }
  ]
  "CurrentMarking" => Symfony\Component\Workflow\Marking {#19
    -places: array:1 [
      "a" => true
    ]
  }
  "can validate t1" => true
  "can validate t3" => false
  "can validate t6" => false
  "apply t1" => Symfony\Component\Workflow\Marking {#22
    -places: array:2 [
      "b" => true
      "c" => true
    ]
  }
  "apply t2" => Symfony\Component\Workflow\Marking {#47
    -places: array:1 [
      "d" => true
    ]
  }
  "can validate t1 bis" => false
  "can validate t3 bis" => true
  "can validate t6 bis" => false
]
```

fabpot committed:
added the first more-or-less working version of the Workflow component

lyrixx committed:
[Workflow] Added initial set of files

Cydonia7 commented on src/Symfony/Component/Workflow/Event/GuardEvent.php:
```diff
@@ -0,0 +1,31 @@
+        return $this->blocked;
+    }
+
+    public function setBlockedd($blocked)
```
There is a typo here: Blockedd -> Blocked

lyrixx commented:
Hi guys. I worked on this component. It's not a [Workflow net](Workflow net) implementation (it's a preti net, with more specifications). But I did not implemented everything like the arc weight and so the ability to set more than one token in a place. I fail to see a valid use case in PHP application.

lyrixx commented:
I fixed remaining `FIXME`. I also build a small demo app but I can not share it right now because I will use it for the Symfony Live Paris on Thursday. I worked like a charm ;)

![screenshot5](https://cloud.githubusercontent.com/assets/408368/14243059/967db7b8-fa53-11e5-9a24-aa56f056b26c.png)

HeahDude commented:
Wow! That looks terrific! Is there any start of documentation for that component ? I can't wait to read some :)

Nicofuma commented on src/Symfony/Component/Workflow/MarkingStore/MarkingStoreInterface.php:
```diff
@@ -0,0 +1,26 @@
+ */
+interface MarkingStoreInterface
+{
+    public function getMarking($subject);
```
could be nice to document a bit the interface methods

lyrixx replied:
Except saying it returns a Marking, I don't know what to say. So for me it's useless.
What do you except?

stloyd replied:
@lyrixx Having phpdoc like:
```php
interface MarkingStoreInterface
{
    /**
     * @return Marking
     */
    public function getMarking($subject);
}
```
Allows IDE's to support this code easier, like highlight etc. cause if you declare that variable contains this interface (not class that implements it) highlight will not work.

Also this is _kinda_ part of Symfony CS, IIRC.

gnugat replied:
@lyrixx you're right, it's pretty straightforward. I suppose some people will complain because their IDE won't guess the type of the return value though.

stof replied:
IDEs cannot guess the return type of an interface, as there is no code in the interface to guess it.

And IDEs guessing it based on implementations would be broken IDEs, as they cannot be sure that the implementation returns all possible return value types (for instance, the return type could be an interface rather than an implementation)

lyrixx replied:
I fixed that anyway ;)

stloyd commented on src/Symfony/Component/Workflow/Dumper/GraphvizDumper.php:
```diff
@@ -0,0 +1,262 @@
+    {
+        $transitions = array();
+
+        foreach ($definition->getTransitions() as $name => $_) {
```
I know it's not used, but IMO this should follow CS & that variable (`$_`) should have normal name ;)

Nicofuma commented on src/Symfony/Component/Workflow/Dumper/DumperInterface.php:
```diff
@@ -0,0 +1,34 @@
+     *
+     * @return string The representation of the workflow
+     */
+    public function dump(Definition $definition, Marking $marking = null, array $options = array());
```
`$marking` is missing in the doc block

Nicofuma commented on src/Symfony/Component/Workflow/Dumper/GraphvizDumper.php:
```diff
@@ -0,0 +1,262 @@
+{
+    private static $defaultOptions = array(
+        'graph' => array('ratio' => 'compress', 'rankdir' => 'LR'),
+        'node' => array('fontsize' => 9, 'fontname' => 'Arial', 'color' => '#333333', 'fillcolor' => 'lightblue', 'fixedsize' => true, 'width' => 1),
```
should be place (according to the doc block of the constructor)

lyrixx replied:
I fixed the constructor doc block.

lyrixx commented:
> Wow! That looks terrific! Is there any start of documentation for that component ? I can't wait to read some :)

@HeahDude: Not yet. I'm a very bad redactor :/ But I'm going to present this component during the SymfonyLive Paris 2016 on Thursday. And I'm going to open source a small SF2 application that use it.

Nicofuma commented on src/Symfony/Component/Workflow/Dumper/GraphvizDumper.php:
```diff
@@ -0,0 +1,262 @@
+    {
+        return sprintf("digraph workflow {\n  %s\n  node [%s];\n  edge [%s];\n\n",
+            $this->addOptions($options['graph']),
+            $this->addOptions($options['node']),
```
should be `place` according to the doc block of the constructor

lyrixx replied:
I fixed the constructor doc block. (`place` does not exist in the graphviz language.)

Nicofuma commented on src/Symfony/Component/Workflow/EventListener/AuditTrailListener.php:
```diff
@@ -0,0 +1,55 @@
+    public static function getSubscribedEvents()
+    {
+        return array(
+            'workflow.leave' => array('onLeave'),
```
Could it make sense to have a final `WorkflowEvents` class (like for HttpKernel, Form or Security)

lyrixx replied:
Yes and no because *many* events name are generated dynamically. See the Workflow class.

Nicofuma replied:
I saw that after yeah...
Actually I don't think it will be that much usage for a such class as in most case the general event is too general (except for logging which is already implemented...)

lyrixx commented:
I fixed the first bunch of comments ;)

stof commented:
@lyrixx I think this is too late for inclusion in 3.1 though (as we are in feature freeze)

lyrixx commented:
> I think this is too late for inclusion in 3.1 though (as we are in feature freeze)

Yes I think so :( 

But the component will be require-able with a symfony 2.X or 3.X thanks to the subtree split.

I'm working on a better integration with the FrameworkBundle (semantic config, auto-wiring).
And I may create a standalone bundle to offer an easy integration for older symfony (like @nicolas-grekas has done with the VarDumperBundle)

liuggio commented:
The Finite State Machine is a good practice, BUT I'm :-1: about creating a component.

What do think about (instead of using an external dependency inside the domain)
to enrich the model moving the Domain Entity with explicit states with explicit methods,
is so easy to implement using value objects, and there's a lot papers about it.

So few simple testable code instead of complex configuration that hide the states..

lyrixx commented:
@liuggio this component is not a state machine, so it is useless to answer your question.

liuggio commented:
@lyrixx I hope you can find time to read what I wrote, I think is adequate.

lyrixx commented:
Yes, I hear you.

But I think I already give my arguments in [the PR description](https://github.com/symfony/symfony/pull/11882) and in [this comment](https://github.com/symfony/symfony/pull/11882#issuecomment-54854864).

---

So let's summary all arguments *once for all*:

* **Symfony is a set of reusable PHP components**. So any component that help end user to write better and simpler code is good to take (Of course only if the use case is valid and "mainstream").
* Workflow Engines exist in many other frameworks (JAVA, .NET, Django, Rails ... ).
* **Utility of Workflow Engine is real**. PetriNet appears in 1962 and it is still used in the industry thanks to a proven model. Moreover there are plenty of State Machine or PetriNet implementations on Packagist. And for some quite high number of downloads (200K for finite). So clearly, there is a need.
* **It brings much more visibility to the Workflow Engine** into the Symfony Community and even the PHP Community.
* **Adding it to Symfony will also enhance the quality / test / documentation** of the component thanks to the contributors community (thanks everyone !!). It will also benefit from the tooling of Symfony: scheduled releases, subtree splits, security patches etc.
* **Too many people (like me) are writing again and again the same boiler code to implement some Sate Machine**. But frankly, for me, it's boring to write it again and again. Internal implementation might contains bugs where Symfony implementation should not (I hope ;)). It's always boring to add a new state with an internal implementation. 
* This component is not a state machine. As you can see in my previous screenshot, **it's an implementation of the Workflow Net** (An extension of PetriNet). So it means a subject can be in different places. (i.e.: different states at the same time). So it allow you, among other things, to parallelize actions.
* **This component implements a Guard system**. For instance, with very very few lines of code, you can block a transition if the user doesn’t have​ the role `ROLE_FOOBAR` (this is an example...). Or anything else you can imagine.
* **This component fire many events that you can listen to**. To build, for example, an audit log/trail of the state of your subjects
* About "domain" code: If you prefer to keep all the logic inside your entity you don't have to use this component. And if you don't use it, it's not an issue. Symfony will not become slower because of a new component. Everything is loaded only when you are using it.
* About "domain" code (part 2): If you put all the logic inside an entity, It will become some spaghetti code because you will need to inject many services to your entity to be able to know if you can apply a transition or not. And this is not the responsibility of an Entity. But I don't want to argue more on this topic because I know many people like to put all there code into their Entites. And I really don't want to debate on this topic here.
* This component is able to **dump the workflow as a graphviz representation**.

egircys commented on src/Symfony/Component/Workflow/Event/Event.php:
```diff
@@ -0,0 +1,49 @@
+        return $this->subject;
+    }
+
+    public function getTransition()
```
Should be plural ```getTransitions()```?

lyrixx replied:
Thanks not noting me this. The issue was on the attributes. So it's  `getTransition` and I renamed the attributes from `transitions` to `transition`.

liuggio commented:
> About "domain" code (part 2): If you put all the logic inside an entity, It will become some spaghetti code

it's called Domain Driven design not spaghetti code.

timglabisch commented on src/Symfony/Component/Workflow/Transition.php:
```diff
@@ -0,0 +1,45 @@
+    public function __construct($name, $froms, $tos)
+    {
+        $this->name = $name;
+        $this->froms = (array) $froms;
```
why not typehint `froms` and `tos` as array? would be much more consistent.

lyrixx replied:
Because it allows you to run `new Transition('the name', 'from', 'to')` instead of `new Transition('the name', ['from'], ['to'])`

lyrixx commented:
I just pushed some code to integrate the component and the Full Stack Framework.

Now, you can use:
```
framework:
    workflows:
        my_workflow:
            marking_store:
                type: property_accessor
            supports:
                - Some\Class
            places:
                - first
                - last
            transitions:
                go:
                    from:
                        - first
                    to:
                        - last
```

and a new service `workflow.my_workflown` ; and some goodness in twig ;)

Oliboy50 commented:
I don't think this component can be merged until it lacks phpdoc for everything.

zemd commented:
it looks very interesting, but lacks a real life examples. when can you share your test app?

zemd commented:
it looks very interesting, but lacks a real life examples. when can you share your test app?

mickaelandrieu commented:
@lyrixx sounds good to me, also this is the first time you take time to realy explain the advantages of this component: thanks for that! You do a great job on this implementation, can't wait to use it in PrestaShop core :+1: 

I only have a comment:

> And if you don't use it, it's not an issue. Symfony will not become slower because of a new component. Everything is loaded only when you are using it.

I'm clearly interested on this field (as you may know ^^), can you provide a blackfire profile for it ?

Because when I was part of the BackBee project, after removing all "useless" dependencies we have noticed better performances (we didnt used blackfire).

Mickaël

stof commented:
> Because when I was part of the BackBee project, after removing all "useless" dependencies we have noticed better performances (we didnt used blackfire).

were they loaded ?

mickaelandrieu commented:
@stof nope, they weren't. I can't explain why but there is an overall. This overall is minimal thanks to the `--optimize` option of composer but still exists.

gnugat commented:
> @mickaelandrieu 
> > @lyrixx 
> > And if you don't use it, it's not an issue. Symfony will not become slower because of a new component. Everything is loaded only when you are using it.
>
> I'm clearly interested on this field (as you may know ^^), can you provide a blackfire profile for it ?
> 
> Because when I was part of the BackBee project, after removing all "useless" dependencies we have noticed better performances (we didnt used blackfire).

I can provide you with some answers :) .

## Autoloading (itself) overhead - negligible

There is a small overhead for each element in a `autoload` section (in all `composer.json` found for your project), but it's negligible: [See comparison profiling graph](https://blackfire.io/profiles/compare/70bcfad5-1028-4b80-8ef5-05d116f0f7be/graph).

> **Note**: Please ignore the execution time (from `0.5ms` to `1.3ms`), as it depends on the computer capacity and the profiling was done on an 8 years old laptop (a [HP Compaq 8510p](http://www.cnet.com/products/hp-compaq-8510p-15-4-core-2-duo-t7700-vista-business-2-gb-ram-120-gb-hdd-series/specs/)).
> We should really be looking on the diff percentage: `+144%` (so ~`x2.5`), for 34 more dependencies. I think you'll agree it's not much.

The above profiling has been done on a PHP application that consist of one line: `require __DIR__.'/vendor/autoload.php';`, with a `composer.json` file and running `composer install -o --no-dev`. The first graph was done with an empty `composer.json`, and the second one was done with a `composer.json` requiring 34 individual Symfony components.

As we can see, `ClassLoader#setPsr4()` is called 34 times more, which is to be expected.

> **Note**: Interestingly `strlen` is also called 34 times more, but since it consists of a function call to read a property it's safe to say that it's negligible.

Using the `symfony/symfony` meta package instead of all its individual components can be regarded as a micro optimization: it does improve things, but not by much.

## Autoloaded dependencies overhead - it depends

Funily enough, just because you don't use directly a dependency doesn't mean it's not used: [See Standard Edition VS Empty Editon VS MicroFrameworkBundle benchmarks](http://gnugat.github.io/micro-framework-bundle/doc/03-benchmark.html).

By having a look at [the first](https://blackfire.io/profiles/compare/2067da45-4605-4f33-beb2-c37a2a9a9740/graph) and [the second](https://blackfire.io/profiles/compare/8097ec35-dd77-4829-9fbc-ac7168231cc9/graph) comparison graphs, we can see that autoloading is quite a big performance eater. By having a closer look, we realize its main caller is `Container#get`, which is mainly called by `ContainerAwareEventDispatcher`.

And that's why adding "unused" dependencies can actually impact performances. Using the Standard Edition without unregistering bundles we don't need will impact performances.

## What about Workflow?

Well, adding Workflow to your project and not using it won't impact significantly your performances. Adding Workflow to the `symfony/symfony` meta package and not using it won't impact your performances at all. Adding Workflow to `FrameworkBundle` will not have any impact on performances if it doesn't add event listeners.

lyrixx commented:
Hi @gnugat;

Even if it's out of topic, I'm going to try to answer you to (I hope) close this
debate once for all.

> We should really be looking on the diff percentage: +144% (so ~x2.5), for 34
> more dependencies. I think you'll agree it's not much.

First, I don't know why you said "It's not so much". Because `+144%` is huge!
But you have to put number in perspective.

Basically, your application does nothing. So it's obvious to me that the
percentage delta means nothing in this case. For example, if you put a
`sleep(10)` in your code, then the percentage delta will be almost equal to 0.
Does it means it has no impact? No. But it's better to understand number. Here,
the absolute delta is `+805 µs`, so it's less than `1ms`. More over, even if we
are trying very hard to reduce the overhead of the Blackfire PHP probe when the
probe profiles a script, the overhead exists. So in real life the autoload
overhead will be much smaller than `1ms`.

So yes indeed, having more file in the `vendor/` directory has an impact. But we
are talking here of something that add less than `1ms` for 1763 new
files (Symfony code base ` find src/ -name '*.php' ! -path "*/vendor/*"  ! -path '*/Tests/*' | wc -l`).
So for real world application it represents something less than 1%.

Now let's come back the Workflow component: It adds 17 classes. So if it's
linear and we are using 1ms for 1763 files (which is really high number), it
will add `0,009` ms. I guess we all agree here to say it's almost nothing.

And finally, We can talk about the footprint of the new integration between the
workflow component and the framework bundle: If you have no workflow registred,
it's exactly the same thing in production. You can easily see it in the diff of
the [FrameworkExtension](https://github.com/symfony/symfony/pull/11882/files
#diff-0e793081ceb720201745c982a568903fR351).

So, adding this component to Symfony or any other component will add less than
`0,009ms` (and this on the basis of values from a 8 year old computer and high
number due to enabled profiling). I guess it's inferior to `1µs` (micro, not
milli !) in the real world.

**So yes, Adding a new component in Symfony (or elsewhere) has not impact on
performance** (as soon as you follow some best practices, do not make me say
what I did not say ;)

lyrixx commented:
Hi @Oliboy50

> it looks very interesting, but lacks a real life examples. when can you share your test app?

Yes, see [this comment](https://github.com/symfony/symfony/pull/11882#issuecomment-205290533)

stloyd commented on src/Symfony/Component/Workflow/MarkingStore/ScalarMarkingStore.php:
```diff
@@ -0,0 +1,48 @@
+    public function __construct($property = 'marking')
+    {
+        $this->property = $property;
+        $this->propertyAccessor = PropertyAccess::createPropertyAccessor();
```
IMO to follow other parts in Symfony this should be optionally injectable:
```php
public function __construct($property = 'marking', PropertyAccessorInterface $propertyAccessor = null)
{
    $this->property = $property;
    $this->propertyAccessor = $propertyAccessor ?: PropertyAccess::createPropertyAccessor();
}

lyrixx replied:
You are right. I fixed it.

stloyd replied:
Seems like you have updated `PropertyAccessorMarkingStore` but not the `ScalarMarkingStore` ;)

lyrixx replied:
Good catch ;) Thanks.

gnugat commented:
The number of files in a package **does not** impact composer autoloading (at least when using the `-o` option), it's the number of `autoload` configurations that does (e.g. in `composer.json`, each elements in the `psr-4` array from the `autoload` section). You can check it by having a look at `vendor/autoload.php` and `vendor/composer/autoload_real.php` in your projects to see how it actually works.

The overhead of composer autoloading increased by +144% when we added +34 composer packages, which means that the overhead of adding a single composer package (e.g. `symfony/workflow`) would be much lower (around +20%, as we can see in [this other test which only adds one dependency](https://blackfire.io/profiles/compare/e0778bf9-33e7-4c27-b525-f3979921cc52/graph)). As @lyrixx pointed out this number is only relative to composer autoloading, not to the application's performances: that's what I mean by "it's not much".

mickaelandrieu commented on src/Symfony/Bundle/FrameworkBundle/Command/WorkflowDumpCommand.php:
```diff
@@ -0,0 +1,67 @@
+ */
+class WorkflowDumpCommand extends ContainerAwareCommand
+{
+    /**
```
can you implement `isEnabled` command to return `false` if you don't add any workflows declared ?

lyrixx replied:
I store nothing in the DIC about existing workflow. So I will not be able to do that with ease. I think I should add a parameter in the DIC for that. 

So my question is: Do you really need this "feature" ?

mickaelandrieu replied:
Can you access configuration from container like you did in the extension class in the Framework extension class ?

Because we already have [a lot of commands](https://github.com/symfony/symfony/issues/15689) and for now, no way to filter except using this function, I think we got a chance to avoid this command when we don't need it.

To sum up:
* if we have a workflow configurated => of course the command should be displayed: great !
* if no workflow configurated for this application => does it make sense to display this command.. probably not.

What do you think ?

lyrixx replied:
Yes ; I agree with you. I will implement that ASAP.

mickaelandrieu replied:
Thank you :)

lyrixx replied:
Fixed, in a better way.

mdwheele commented:
@lyrixx I really appreciate the work here. I am at a point in a current project of mine where I'm looking to replace a partial WF-net implementation with a third-party component. 

In order to fulfill workflow client application and application invocation APIs to implement work item brokering (or *resource in-basket*, as it is commonly referred), I need to be able to know when a transition has become enabled; thus signaling a task is available for work. It seems with the current implementation, I do have "on.enter" events and could check for myself whether or not impacting transitions are enabled, but it would be very slick (in my opinion) if this were just another event that I could subscribe to. 

Would you consider adding an additional event for when transitions become enabled before features for `symfony/workflow` are frozen? I completely understand if you want to limit scope as there looks to be a good bit of work ahead of you already in documentation. 

Thanks!

lyrixx commented:
Hello @mdwheele 

Thanks for your comment.

So, I will rephrase your comment, just to be sure we are talking about the same thing.

You want a new event, that will occurs after the transition, to inform you the new available transitions ?

If so, you can already achieve this by youserlf: You listen all transition, and you call `$this->workflow->getAvailableTransitions()`.

mdwheele commented:
> You want a new event, that will occurs after the transition, to inform you the new available transitions ?

That's correct. I understand I could listen for fire events and call `$this->workflow->getAvailableTransitions()` (though I would have expected to find that feature by `getEnabledTransitions()`). I would then need to maintain a thicker abstraction between this Component and a sub-system responsible for delegating work items (enabled transitions) to resources. That abstraction would then be forced to be responsible for listening for fire events, querying available transitions and publishing diffs in that state to subscribers in a "work item brokering" domain. For that reason, it is useful (to at least me) to have the event API be a little more verbose even if there are work-arounds to query the component statefully.

Also, I don't mean to interrupt and can hold this until everything is merged if that is preferable. I understand this PR is about working out last-issues with merging into the framework.

lyrixx commented:
Actually, your use case seems good and I think I can implement that directly in the core. Right now I have to polish everything, but I keep your feature in mind. 

And about `getAvailableTransitions` vs `getEnabledTransitions` it's a very fair point. I may renamed it too.

mikeSimonson commented on src/Symfony/Component/Workflow/Dumper/GraphvizDumper.php:
```diff
@@ -0,0 +1,263 @@
+    {
+        $transitions = array();
+
+        foreach ($definition->getTransitions() as $name => $nbToken) {
```
What do you mean by $nbToken ?

lyrixx replied:
I do not use this value. But this is the vocabulary of petrinet implementation. I'm sure [wikipedia](https://en.wikipedia.org/wiki/Petri_net#Petri_net_basics) would have a much better explanation than me ;)

javiereguiluz replied:
@lyrixx maybe we could rename `$nbToken` to `$numToken` or `$numTokens` to make it easier to understand.

lyrixx replied:
To me, `numToken` means the "token number", or "token numéro". where here is the total number of token. All that said, I'm not a native English (obviously ;)). So what should I do ?

hacfi replied:
Maybe I’m missing something here but don’t we just get an instance of `Symfony\Component\Workflow\Transition` here? In that case why not just use ` $transition` like in https://github.com/symfony/symfony/pull/11882/files#diff-16ec272db47080b27d1118e196175dc9R153?

nicolassing replied:
@lyrixx for counters I usually name them $xxxCount. $tokenCount in your case :)

lyrixx replied:
Actually, it was a typo. it should be transition. It's not fixed.

nicolas-grekas commented:
@mickaelandrieu @gnugat I don't want to hijack the thread but since there are some discussions already about the overhead introduced in the autoloader by non-used classes, here is my would-like-to-be-definitive-answer to this "overhead" non-sense: https://github.com/composer/composer/pull/5172

mickaelandrieu commented:
@nicolas-grekas thank you, as always, you're great! :+1:

gnugat commented:
@nicolas-grekas: I now realize that I might not have been talking about the same overhead as you both. My profilings were aimed at the ClassLoader initialization, which happens even if no classes are actually loaded. Your PR seems to target the actual class loading, which happens for each class needed to be loaded.

nicolas-grekas commented:
@gnugat I guess you should read my PR again then...

lyrixx commented:
Hello.

I open sourced my little demo app:

* code: https://github.com/lyrixx/SFLive-Paris2016-Workflow
* demo: http://symfony-workflow-demo.herokuapp.com/show/1
* slide (in French, for the SymfonyLive Paris 2016 ; I might translate them): https://speakerdeck.com/lyrixx/le-reveil-du-workflow

lyrixx commented:
Hello @mdwheele ; I implemented your feature request ;)

StephenOTT commented:
Has Anyone looked at what OroPlatform / OroBAP is doing with workflow?
https://github.com/orocrm/platform/tree/master/src/Oro/Bundle/WorkflowBundle
Two examples:
![screen shot 2016-04-13 at 7 09 07 pm](https://cloud.githubusercontent.com/assets/1994838/14512234/4591da64-01ab-11e6-98bf-fb9e73441c7e.png)
![screen shot 2016-04-13 at 7 08 34 pm](https://cloud.githubusercontent.com/assets/1994838/14512233/4591990a-01ab-11e6-959a-03eb0bcd1681.png)

StephenOTT commented:
I have gone through the writeup for this PR a few times and I am little confused based on the back and forth.  The goal is to create a workflow around state machine or to create workflow independent of state machine?

Guess what i am asking is are you creating a structured order to state machine or are you creating flowchart workflow?
I love wiki's explanation and diagram: https://en.wikipedia.org/wiki/State_diagram#State_diagrams_versus_flowcharts
Are you trying to do A or B?
![statechart_vs_flowchart-1](https://cloud.githubusercontent.com/assets/1994838/14512297/c2e6f332-01ab-11e6-8a6c-0a134005f037.png)

lyrixx commented:
> . The goal is to create a workflow around state machine or to create workflow independent of state machine?

The goal is to implement something to manage object states. And Actually, it's an implementation of ["workflow net"](http://martinfowler.workflowpatterns.com/documentation/documents/vanderaalst98application.pdf) ; So it's not a (finite) state machine.

> Guess what i am asking is are you creating a structured order to state machine or are you creating flowchart workflow?

Because it's not a state machine, I cannot answer this question with accurateness. But basically, in order to move an object in the workflow, you have to call some code (`apply($subject, $transitionName)`). And nothing is automatic. I mean a transition enabled is not automatically fired. Everything should be explicit

mdwheele commented on src/Symfony/Component/Workflow/Dumper/GraphvizDumper.php:
```diff
@@ -0,0 +1,263 @@
+    {
+        $code = '';
+
+        foreach ($transitions as $id => $place) {
```
Small typo: `$place` should be `$transition`

HeahDude replied:
@lyrixx don't forget this one :)

mdwheele commented on src/Symfony/Component/Workflow/Dumper/GraphvizDumper.php:
```diff
@@ -0,0 +1,263 @@
+    {
+        $code = '';
+
+        foreach ($edges as$edge) {
```
Missing space between `as` and `$edge`.

mikeSimonson commented:
Hi @lyrixx,

First thank you very much for that nice component.
Before it goes gold, I have a few suggestion that I'd like your input on and that I would happily implement.

By toying a little bit with the code, I realised that some classes like Definition and Transition were lacking some doc blocks and while I was adding those I noticed that the Places are not an object but only a string.

This seemed odd to me and I wanted to know what you though about making Place an object and then adding those docblocks ?

Regards

lyrixx commented:
Hello @mikeSimonson 

Thanks for your feedback.

I'm not against your change, but would be the benefit of using a class instead of a string?

mikeSimonson commented:
@lyrixx It makes a lot of the code that use those concept much more expressive.
Likewise the return type (docblock) of a bunch of method becomes less confusing ( I am getting back a Place not a string that might or might not represent a Place) etc.

ro0NL commented:
Just found a petrinet library on github, started 4 days ago :laughing: 
https://github.com/ziguss/petrinet

Maybe someone can compare?

VaN-dev commented:
If I understand how this component works, a marking has to own all the "froms" places of a transition for this transition to be relevant ?

For example, I have this transition :

lyrixx commented:
Hello @VaN-dev 

> If I want to apply this "review" transition to an entity, this entity has to be in both "draft" AND "discarded" states ?

yes, indeed.

> Or is there some kind of "OR" feature (without having to duplicate this transition for every "from" possible) ?

No, you have to create several transition for erevy form

mikeSimonson commented:
@lyrixx Here is the PR that we talked about https://github.com/lyrixx/symfony/pull/8

Any comments ? Thanks

mikeSimonson commented:
@lyrixx ping

lyrixx commented:
Hello @mikeSimonson 

I was on holidays, that's why I did not reply you as fast as you expected.

Anyway Thanks for your works ; 

Some feedback on your work:

* I don't like PHPDoc, but because you already added it and I don't like wasting work, I'm going to cherry-pick them
* you added code like `Place::fromNames` but this is dead code (only used in tests classe).
* I still don't understand why a `Place` object is better than a simple string ; now the code is harder to write and lead to WTF (ex: why do you have twice `place(a)`:
![screenshot1](https://cloud.githubusercontent.com/assets/408368/15141955/fd7866f0-16a2-11e6-808c-577d2fd93427.png)

mikeSimonson commented:
Hi,

I hope you had nice holidays.

> I don't like PHPDoc, but because you already added it and I don't like wasting work, I'm going to cherry-pick them

The phpdoc is there because for code that is pre php7 it allow the ide to provide contextual information.

> you added code like Place::fromNames but this is dead code (only used in tests classe).

The  ```Place::fromNames``` is there so that when you need to create a lot of Place from an array of string you have a handy shortcut to do it. I planned to add xml configuration (again for autocomplete and autocheck) where it would be useful.

> I still don't understand why a Place object is better than a simple string ; now the code is harder to write : 

The Place object is useful because it convey meaning in the context of a Petry net implementation. You are not returning a string, you return a Place. It is much more easier to understand the code when you pass around object that have a meaning in the context of the Petry net then when you pass a string and it's anyone guess what the string will represent this time.
Thanks to that you also get for free type checking and autocomplete in the ide, but type checking and contextual information are really the biggest benefit.
It also make little sense to use an object for the transition and then a string for a place. Then why not using array everywhere? => because it's harder to understand and debug.

> and lead to WTF (ex: why do you have twice place(a)

I just wanted to verify in the test here that it wasn't going to cause a problem with two different instances of the same Place.

I also changed the Transition and Definition so that you cannot instantiate invalid object ( aka required parameter to have a valid object are in the constructor ).

Regards

jeromeWeb commented:
Hi,
Is there a targeted version of symfony for this component ?

lyrixx commented:
Hello @jeromeWeb 

> Is there a targeted version of symfony for this component ?

Yes, it's 2.2

jeromeWeb commented:
@lyrixx, ok cool !

( Version 2.2 I was thinking something like ... 😕 ❓ missed ! )

welcoMattic commented:
Hello @lyrixx,

Pending 3.2, do you plan to release a bundle?

mikeSimonson commented:
@lyrixx If you want a better explanation of my PR you can ask @stof to review it maybe you know him.

StephenOTT commented:
@lyrixx have you thought about how this could be used to control Page Flow of Symfony Forms? (Such as wizards, or complex multi-page interactions).

Also, is there support for transitioning to previous "places".  Would this just be a reversed transition?

lyrixx commented:
@mikeSimonson I integrated your work in this PR. I also squashed yours commits and reverted few things. I'm gonna comment in your PR to not pollute this thread.

dunglas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -345,6 +347,54 @@ private function registerProfilerConfiguration(array $config, ContainerBuilder $
+        $registryDefintion = $container->getDefinition('workflow.registry');
+
+        foreach ($workflows as $name => $workflow) {
+            $definitionDefinition = new Definition('Symfony\Component\Workflow\Definition');
```
You can use `Definition::class` instead.

dunglas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -345,6 +347,54 @@ private function registerProfilerConfiguration(array $config, ContainerBuilder $
+            $definitionDefinition = new Definition('Symfony\Component\Workflow\Definition');
+            $definitionDefinition->addMethodCall('addPlaces', [$workflow['places']]);
+            foreach ($workflow['transitions'] as $transitionName => $transition) {
+                $definitionDefinition->addMethodCall('addTransition', [new Definition('Symfony\Component\Workflow\Transition', [$transitionName, $transition['from'], $transition['to']])]);
```
You can use the `::class` constant here too.

dunglas commented on src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/php/workflow.php:
```diff
@@ -0,0 +1,28 @@
+                'type' => 'property_accessor',
+            ),
+            'supports' => array(
+                'Symfony\Bundle\FrameworkBundle\Tests\DependencyInjection\FrameworkExtensionTest',
```
`:class`

dunglas replied:
`:class`

dunglas commented on src/Symfony/Component/Workflow/Dumper/GraphvizDumper.php:
```diff
@@ -0,0 +1,265 @@
+
+        foreach ($edges as$edge) {
+            $code .= sprintf("  %s_%s -> %s_%s [style=\"solid\"];\n",
+                $edge['direction'] === 'from' ? 'place' : 'transition',
```
Should follow Yoda style (same line 188).

dunglas commented on src/Symfony/Component/Workflow/MarkingStore/PropertyAccessorMarkingStore.php:
```diff
@@ -0,0 +1,65 @@
+    private $property;
+
+    /**
+     * @var PropertyAccessorInterface
```
This dockbloc is useless.

xabbuh replied:
I agree, we don't add docblocks for properties whose type can be inferred by IDEs from the constructor.

lyrixx replied:
:+1: I hate almost all docblock. It was not added by me but it's with great pleasure I'm going to remove it ;)

dunglas commented:
`LICENSE`, `README.md` and `CHANGELOG.md` are missing. Except minor comments I left, 👍 for me.

fabpot commented on src/Symfony/Component/Workflow/Event/GuardEvent.php:
```diff
@@ -0,0 +1,31 @@
+
+    public function setBlocked($blocked)
+    {
+        $this->blocked = (Boolean) $blocked;
```
bool

mauchede commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -345,6 +347,54 @@ private function registerProfilerConfiguration(array $config, ContainerBuilder $
+
+        foreach ($workflows as $name => $workflow) {
+            $definitionDefinition = new Definition('Symfony\Component\Workflow\Definition');
+            $definitionDefinition->addMethodCall('addPlaces', [$workflow['places']]);
```
[Constructor of Definition](https://github.com/lyrixx/symfony/blob/component-workflow/src/Symfony/Component/Workflow/Definition.php#L43) requires two arguments (`places` and `transitions`). The service will be invalid.

mauchede replied:
Method `addPlaces` uses the method [addPlace](https://github.com/lyrixx/symfony/blob/component-workflow/src/Symfony/Component/Workflow/Definition.php#L88). This method requires a `Place` instance, but you will call it with an array of strings.

lyrixx replied:
I know. This part is not updated yet. But anyway I'm going to revert the last change about the Place object.

mauchede commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -345,6 +347,54 @@ private function registerProfilerConfiguration(array $config, ContainerBuilder $
+
+        $loader->load('workflow.xml');
+
+        $registryDefintion = $container->getDefinition('workflow.registry');
```
`registryDefintion` -> `registryDefinition`

lyrixx commented:
@fabpot Hello ! 

I worked a bit on this PR:

* I reverted the introduction of the `Place` class + fixed tests
* I fixed pending comments
* I updated https://github.com/lyrixx/SFLive-Paris2016-Workflow
* I updated http://symfony-workflow-demo.herokuapp.com
* I played a bit with it.

It's almost ready, but it needs another full review.
And I let all PHPDoc added by @mikeSimonson. IMHO, there are too many,
but as there are now here ... But I will be so happy to remove them, just tell me.

HeahDude commented on src/Symfony/Bridge/Twig/Extension/WorkflowExtension.php:
```diff
@@ -0,0 +1,52 @@
+    public function getFunctions()
+    {
+        return array(
+            new \Twig_SimpleFunction('workflow_can', [$this, 'canTransition']),
```
Can we use the short array syntax for this component sf3 only or should it remain consistent with the codebase?

lyrixx replied:
thanks, fixed.

HeahDude commented on src/Symfony/Bundle/FrameworkBundle/Command/WorkflowDumpCommand.php:
```diff
@@ -0,0 +1,79 @@
+            ))
+            ->setDescription('Dump a workflow')
+            ->setHelp(<<<'EOF'
+The <info>%command.name%</info> command dump an dot representation of a workflow.
```
typo `a dot`?

fabpot commented:
@lyrixx Please, remove any phpdocs that are not useful or do not bring any value.

chalasr commented on src/Symfony/Component/Workflow/Dumper/DumperInterface.php:
```diff
@@ -0,0 +1,35 @@
+     * Dumps a workflow definition.
+     *
+     * @param Definition $definition A Definition instance
+     * @param Marking    $marking    A Marking instance
```
Should be `Marking|null`

chalasr commented on src/Symfony/Component/Workflow/Dumper/GraphvizDumper.php:
```diff
@@ -0,0 +1,208 @@
+     * Dumps the workflow as a graphviz graph.
+     *
+     * Available options:
+     *
```
What about `{@inheritdoc}`?

lyrixx replied:
1. It's done just bellow
2. Because theses options are relatives to **this** dumper.

chalasr replied:
The 1 was not here when I commented.

lyrixx commented:
@fabpot done ;)

javiereguiluz commented on src/Symfony/Bundle/FrameworkBundle/Command/WorkflowDumpCommand.php:
```diff
@@ -0,0 +1,79 @@
+            ->setName('workflow:dump')
+            ->setDefinition(array(
+                new InputArgument('name', InputArgument::REQUIRED, 'A workflow name'),
+                new InputArgument('marking', InputArgument::IS_ARRAY, 'A marking (a list of place)'),
```
`a list of place` -> `a list of places`

javiereguiluz commented on src/Symfony/Bundle/FrameworkBundle/Command/WorkflowDumpCommand.php:
```diff
@@ -0,0 +1,79 @@
+use Symfony\Component\Workflow\Marking;
+
+/**
+ * A console command for retrieving information about services.
```
Is this class description correct? It doesn't even mention the workflows:

```
A console command for retrieving information about services.
```

javiereguiluz commented on src/Symfony/Bridge/Twig/Extension/WorkflowExtension.php:
```diff
@@ -0,0 +1,52 @@
+    {
+        return array(
+            new \Twig_SimpleFunction('workflow_can', array($this, 'canTransition')),
+            new \Twig_SimpleFunction('workflow_enabled_transitions', array($this, 'getEnabledTransitions')),
```
Could we rename `workflow_enabled_transitions` by `workflow_transitions` ... or would that be confusing?

javiereguiluz commented on src/Symfony/Bundle/FrameworkBundle/Command/WorkflowDumpCommand.php:
```diff
@@ -0,0 +1,79 @@
+            ))
+            ->setDescription('Dump a workflow')
+            ->setHelp(<<<'EOF'
+The <info>%command.name%</info> command dump a dot representation of a workflow.
```
`dump` -> `dumps`

javiereguiluz replied:
`a dot representation of a workflow` is hard to understand unless you know what a `DOT` filetype is or what Graphviz is. What about something like:

```
... dumps the graphical representation of a workflow in DOT format.
```

javiereguiluz commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -226,6 +227,99 @@ private function addProfilerSection(ArrayNodeDefinition $rootNode)
+                                ->isRequired()
+                                ->children()
+                                    ->enumNode('type')
+                                        ->values(['property_accessor', 'scalar'])
```
The code mixes the short array notation and the traditional array notation. IIRC, only the traditional notation is allowed, even for modern Symfony versions.

lyrixx commented:
@javiereguiluz Thanks for the review. I addressed your comments.

HeahDude commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -347,6 +350,54 @@ private function registerProfilerConfiguration(array $config, ContainerBuilder $
+            $definitionDefinition = new Definition(Workflow\Definition::class);
+            $definitionDefinition->addMethodCall('addPlaces', [$workflow['places']]);
+            foreach ($workflow['transitions'] as $transitionName => $transition) {
+                $definitionDefinition->addMethodCall('addTransition', [new Definition(Workflow\Transition::class, [$transitionName, $transition['from'], $transition['to']])]);
```
`array(..)`

HeahDude commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -347,6 +350,54 @@ private function registerProfilerConfiguration(array $config, ContainerBuilder $
+            $container->setDefinition($workflowId, $workflowDefinition);
+
+            foreach ($workflow['supports'] as $supportedClass) {
+                $registryDefinition->addMethodCall('add', [new Reference($workflowId), $supportedClass]);
```
`array(..)`

lyrixx replied:
Thanks. Fixed (both)

HeahDude commented on src/Symfony/Component/Workflow/CHANGELOG.md:
```diff
@@ -0,0 +1,2 @@
+CHANGELOG
+=========
```
Why not:
```md
3.2.0
-----

 * added the component
```
?

lyrixx replied:
IIRC it's not done for other component.

HeahDude commented on src/Symfony/Component/Workflow/Definition.php:
```diff
@@ -0,0 +1,109 @@
+    public function addPlace($place)
+    {
+        if (!preg_match('{^[\w\d_-]+$}', $place)) {
+            throw new InvalidArgumentException(sprintf('The place "%s" contains invalid characters.', $name));
```
missing `\` or use statement?

gnugat replied:
There's already a use statement (line 14, `use Symfony\Component\Workflow\Exception\InvalidArgumentException;`)

HeahDude replied:
@gnugat it was not when I commented ;)

lyrixx replied:
@HeahDude Indeed. Good catch. Thanks.

parweb replied:
@gnugat @lyrixx is it fix ? because, in my project it's not working today.

my class is App\Customer and i have an InvalidArgumentException

It doesn't like the backslash in the namespace

lyrixx replied:
Hello @parweb 

Could you open a new issue?

parweb replied:
The issue wasn't your code, it's me ^^
I believed the __construct method of Definition class, was for giving a name to the workflow.
But it's just an another way to add places.

lyrixx replied:
Yes, you can add places via the constructor or via this method. Both do the same things.

HeahDude commented on src/Symfony/Component/Workflow/Dumper/GraphvizDumper.php:
```diff
@@ -0,0 +1,204 @@
+     *  * node: The default options for nodes (places + transitions)
+     *  * edge: The default options for edges
+     *
+     * {@inheritdoc}
```
Isn't `{@inheritdoc}` usually before specific doc?

HeahDude commented on src/Symfony/Component/Workflow/Dumper/GraphvizDumper.php:
```diff
@@ -0,0 +1,204 @@
+        $places = array();
+
+        foreach ($definition->getPlaces() as $place) {
+            $attributes = [];
```
`array()`

dosten replied:
Since this PR is going to be merged into 3.2 which requires PHP >=5.5 we can use the short syntax.

HeahDude commented on src/Symfony/Component/Workflow/Dumper/GraphvizDumper.php:
```diff
@@ -0,0 +1,204 @@
+    private function findPlaces(Definition $definition, Marking $marking = null)
+    {
+        $places = array();
+
```
This empty line can be removed for consistency

lyrixx replied:
With what ? You added 4 times the same comments, so I thinks this "inconsistency" is consistent. And Anyway, I prefer like that ;)

HeahDude commented on src/Symfony/Component/Workflow/Dumper/GraphvizDumper.php:
```diff
@@ -0,0 +1,204 @@
+    private function findTransitions(Definition $definition)
+    {
+        $transitions = array();
+
```
same here

lyrixx replied:
idem

HeahDude commented on src/Symfony/Component/Workflow/Dumper/GraphvizDumper.php:
```diff
@@ -0,0 +1,204 @@
+    private function addPlaces(array $places)
+    {
+        $code = '';
+
```
same here

lyrixx replied:
idem

HeahDude commented on src/Symfony/Component/Workflow/Dumper/GraphvizDumper.php:
```diff
@@ -0,0 +1,204 @@
+    private function addTransitions(array $transitions)
+    {
+        $code = '';
+
```
idem

lyrixx replied:
same here

HeahDude commented on src/Symfony/Component/Workflow/MarkingStore/MarkingStoreInterface.php:
```diff
@@ -0,0 +1,39 @@
+interface MarkingStoreInterface
+{
+    /**
+     * Get a Marking from a subject.
```
`Gets`? if yes, same goes for `Sets` below.

HeahDude commented on src/Symfony/Component/Workflow/MarkingStore/PropertyAccessorMarkingStore.php:
```diff
@@ -0,0 +1,56 @@
+     */
+    public function getMarking($subject)
+    {
+        return new Marking($this->propertyAccessor->getValue($subject, $this->property) ?: []);
```
`array()`

jderusse commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -226,6 +227,99 @@ private function addProfilerSection(ArrayNodeDefinition $rootNode)
+                                    ->arrayNode('arguments')
+                                        ->beforeNormalization()
+                                            ->ifString()
+                                            ->then(function ($v) { return [$v]; })
```
`array($v)`

jderusse commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -226,6 +227,99 @@ private function addProfilerSection(ArrayNodeDefinition $rootNode)
+                                ->isRequired()
+                                ->beforeNormalization()
+                                    ->ifString()
+                                    ->then(function ($v) { return [$v]; })
```
same here

jderusse commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -347,6 +350,54 @@ private function registerProfilerConfiguration(array $config, ContainerBuilder $
+
+        foreach ($workflows as $name => $workflow) {
+            $definitionDefinition = new Definition(Workflow\Definition::class);
+            $definitionDefinition->addMethodCall('addPlaces', [$workflow['places']]);
```
`array(`

HeahDude commented on src/Symfony/Component/Workflow/Registry.php:
```diff
@@ -0,0 +1,65 @@
+
+        foreach ($this->workflows as list($workflow, $classname)) {
+            if ($this->supports($workflow, $classname, $subject, $workflowName)) {
+                if ($matched) {
```
When is this `true`?

lyrixx replied:
never. But it can be equals to a `Workflow` instance.

jderusse commented on src/Symfony/Component/Workflow/Tests/DefinitionTest.php:
```diff
@@ -0,0 +1,73 @@
+        $places = range('a', 'b');
+
+        $transition = new Transition('name', $places[0], $places[1]);
+        $definition = new Definition($places, [$transition]);
```
`array`

jderusse commented on src/Symfony/Component/Workflow/Tests/DefinitionTest.php:
```diff
@@ -0,0 +1,73 @@
+    {
+        $places = range('a', 'b');
+
+        new Definition($places, [new Transition('name', 'c', $places[1])]);
```
same here and bellow

jderusse commented on src/Symfony/Component/Workflow/Tests/Dumper/GraphvizDumperTest.php:
```diff
@@ -0,0 +1,203 @@
+
+    public function provideWorkflowDefinitionWithMarking()
+    {
+        yield [
```
`array` (many times in this file)

stof commented:
@fabpot could you enable the LongArraySyntaxFixer contrib fixer on fabbot.io for the Symfony repo, so that enforcing consistent usage of the long array syntax is done by the tool rather than by the reviewers ?

HeahDude commented on src/Symfony/Component/Workflow/Tests/EventListener/AuditTrailListenerTest.php:
```diff
@@ -0,0 +1,54 @@
+{
+    public function testItWorks()
+    {
+        $transitions = [
```
`array(...)`

HeahDude commented on src/Symfony/Component/Workflow/Tests/EventListener/AuditTrailListenerTest.php:
```diff
@@ -0,0 +1,54 @@
+            new Transition('t2', 'a', 'b'),
+        ];
+
+        $definition = new Definition(['a', 'b'], $transitions);
```
`array(...)`

HeahDude commented on src/Symfony/Component/Workflow/Tests/EventListener/AuditTrailListenerTest.php:
```diff
@@ -0,0 +1,54 @@
+
+        $workflow->apply($object, 't1');
+
+        $expected = [
```
`array(...)`

stof replied:
@HeahDude please don't comment for this everywhere. fabbot.io will handle it on next push to the PR

HeahDude commented on src/Symfony/Component/Workflow/Tests/MarkingStore/PropertyAccessorMarkingStoreTest.php:
```diff
@@ -0,0 +1,32 @@
+
+        $markingStore->setMarking($subject, $marking);
+
+        $this->assertSame(['first_place' => 1], $subject->myMarks);
```
`array(...)`

HeahDude commented on src/Symfony/Component/Workflow/Tests/MarkingTest.php:
```diff
@@ -0,0 +1,35 @@
+
+        $this->assertTrue($marking->has('a'));
+        $this->assertFalse($marking->has('b'));
+        $this->assertSame(['a' => 1], $marking->getPlaces());
```
`array(...)`

HeahDude commented on src/Symfony/Component/Workflow/Tests/MarkingTest.php:
```diff
@@ -0,0 +1,35 @@
+
+        $this->assertTrue($marking->has('a'));
+        $this->assertTrue($marking->has('b'));
+        $this->assertSame(['a' => 1, 'b' => 1], $marking->getPlaces());
```
`array(...)`

fabpot commented:
It has been enabled yesterday

fabpot commented:
ah, no, I enabled it on Symfony, but I forgot to deploy the same on fabbot.io. Fixed now

stof commented:
@fabpot can you re-trigger the inspection for this PR as it contains many usages of the short array syntax ?

fabpot commented:
Done

jderusse commented on src/Symfony/Component/Workflow/EventListener/AuditTrailListener.php:
```diff
@@ -0,0 +1,55 @@
+{
+    public function __construct(LoggerInterface $logger)
+    {
+        $this->logger = $logger;
```
variable is not declared

lyrixx commented:
I think I have fixed all reported issues.

fabpot commented:
Thank you @lyrixx.

robinduval commented:
Happy merge @lyrixx

parweb commented:
@lyrixx , you did a great job.

I implemented the component in my project, but I don't understand something.

The "announce" event ? 

Can you explain a bit what's the propose of this ?
Or a use case ?

guard ----> leave ----> transition ----> enter
those events i see clearly what can i do at what moment but announce it's not clear.
Because it's about the next state.

lyrixx commented:
Hello @parweb 

The `announce` event is here to fit some needs asked in [this comment](https://github.com/symfony/symfony/pull/11882#issuecomment-206411287). If it's still not clear for you, don't hesitate to ask ;)

parweb commented:
@lyrixx i think i understand the purpose better now.

When a new transition is available a event is fire ?

lyrixx commented:
Yes.

tacman commented:
Is there any other documentation besides the first post in this issue?  Is that documentation still current?  I've been using the demo app code (https://github.com/lyrixx/SFLive-Paris2016-Workflow), but now need to dynamically create a workflow instead of coding it in the config file.  Thanks.

hacfi commented:
@tacman Afaik not yet! The goal is to have the docs ready when 3.2 will be released which according to http://symfony.com/doc/current/contributing/community/releases.html is November 2016.

cravler commented:
Is this doable?

![workfolw](https://cloud.githubusercontent.com/assets/157440/17092404/184eb360-524b-11e6-97df-721bbbe5971f.png)

lyrixx commented:
yes

cravler commented:
Example?

lyrixx commented:
@cravler Could you, please, be polite and make real sentences. I'm not a bot.

```yaml
places: [a, b, c, d]
transitions:
  t1: { from: a , to: [b, c]}
  t2: { from: [b, c] , to: d}
```

cravler commented:
Step 1
![workfolw1](https://cloud.githubusercontent.com/assets/157440/17099943/21113364-5274-11e6-932e-f8e9c1587b06.png)
Step 2
![workflow2](https://cloud.githubusercontent.com/assets/157440/17099945/269c895a-5274-11e6-8420-04aa903a1013.png)
Step 3
![workflow3](https://cloud.githubusercontent.com/assets/157440/17099949/28d88958-5274-11e6-8fd6-537dfc75f246.png)
Step 4
![workflow4](https://cloud.githubusercontent.com/assets/157440/17099950/2b79cbfe-5274-11e6-9413-4005f576e742.png)

mdwheele commented:
@cravler You should really add more to your comments so intent is clear. I have extra time, so I'll guess. 

If your intent is that the order in which `paid` and `sent` are marked matters, then you should be using some form of [sequential processing](http://www.workflowpatterns.com/patterns/control/basic/wcp1.php). In this case, when `t1` fires, both `paid` and `sent` should be immediately marked. In fact, there is no "Step 2" and "Step 3". "Step 1" is accurate. The next marking would be something similar to your "Step 3" except `packed` should not be marked (which is what I'm assuming `green` means). The next marking after `t2` fires would be that `delivered` be marked with no other places marked.

You probably want something like:

(pseudo-markup; unsure if this is valid configuration for workflow component)
```yaml
places: [packed, paid, sent, delivered]
transitions:
  t1: { from: packed, to: paid}
  t2: { from: paid, to: sent}
  t3: { from: sent, to: delivered}
```

Go check out http://www.workflowpatterns.com. It is a valuable resource for learning how these data structures work and gives you some hints on what patterns exist and what they're good for. Most should be able to be implemented using this component from what I have seen.

mdwheele commented:
> When a new transition is available an event is fire ?

@parweb That's exactly correct! The immediate use-case I had for this at the time was integration of a workflow with a work item delegator as part of a workflow management system. In the language of Workflow Nets, "places" are called "conditions" and "transitions" are called "tasks". Some tasks are "user / input triggered" meaning that they do not fire immediately upon becoming enabled. An enabled task becomes a "work item" available for work applied by some resource. This is why it is handy to know *when* a transition becomes enabled; so that a collaborator can delegate responsibility for a work item to one or more candidate resources. Once the work item is completed (outside the scope of this component) then the task (transition) will be fired and the Petri net goes on functioning as per usual.

cravler commented:
I asked about this: [Interleaved Parallel Routing](http://www.workflowpatterns.com/patterns/control/state/wcp17_animation.php)

mdwheele commented:
> I asked about this: Interleaved Parallel Routing

I believe, you could implement [Interleaved Parallel Routing (implementation guide)](http://www.workflowpatterns.com/patterns/control/state/wcp17.php) with this component. You have guards / guard events to be able to hook into the execution of the petri net. However, you'll need to implement something to carry the state required to implement the rest, I imagine. 

Keep in mind, this package implements a Petri net first. Many workflow abstractions are able to be projected to a lower-level PT-net form, but will require a bit of orchestration around primitives. The linked document above shows a PT-net implementation of the graphic you linked. It is clearly not as graceful, but the graphic linked was also an abstraction used for demonstrative purposes.

lutangar commented on src/Symfony/Component/Workflow/Workflow.php:
```diff
@@ -0,0 +1,274 @@
+
+        foreach ($this->definition->getTransitions() as $transition) {
+            if ($this->doCan($subject, $marking, $transition)) {
+                $this->dispatcher->dispatch(sprintf('workflow.%s.announce.%s', $this->name, $transition->getName()), $event);
```
@lyrixx @mdwheele I might be wrong, but I was expecting the dispatched event to hold the announced transition instead of the initial transition. 
From where the availabilty of the transition originated from, might not be what's matter the most here. 
After all, the use case was to announce *which transition are now available*.

lyrixx replied:
Hello @lutangar ;

I'm not sure to understand what you want. Anyway, If you want a new feature, or to change an existing one, please open a new issue, it will be easier to discuss about it there. And feel free to ping me ;)

javiereguiluz commented:
I'm locking this pull request because it's generating too many messages:

* If you want to report an error or ask for a new feature, please create a new issue.
* If you are looking for documentation, there is still none but we're discussing about it [in this issue](https://github.com/symfony/symfony-docs/issues/6677). Alternatively, there are some useful resources:
  * [New in Symfony 3.2: Workflow component](http://symfony.com/blog/new-in-symfony-3-2-workflow-component) blog post.
  * [Demo application showcasing the component](https://github.com/lyrixx/SFLive-Paris2016-Workflow)
  * *[Le reveil du Workflow](https://speakerdeck.com/lyrixx/le-reveil-du-workflow)* a presentation in French unveiling the component and showcasing its features.

Thank you all for this long discussion!
